third_party/amd/lib/TritonAMDGPUTransforms/Utility.cpp

```diff
@@ -0,0 +1,96 @@
+#include "Utility.h"
+
+#include "mlir/Dialect/SCF/IR/SCF.h"
+
+namespace findMin {
+int findMinInBlock(Block &block,
+                   const std::function<int(Operation *)> &countFunc);
+
+// Returns the minimum found when accumulating countFunc(op) for all paths
+// between op1 and op2
+int findMinBetweenOps(Operation *op1, Operation *op2,
+                      const std::function<int(Operation *)> &countFunc) {
+  assert(op1 && op2);
+  int count = 0;
+  for (auto op = op1; op != op2; op = op->getNextNode()) {
+    if (auto ifOp = llvm::dyn_cast<scf::IfOp>(op)) {
+      assert(!ifOp.getThenRegion().empty() && !ifOp.getElseRegion().empty());
+      auto minThen = findMinInBlock(ifOp.getThenRegion().front(), countFunc);
+      auto minElse = findMinInBlock(ifOp.getElseRegion().front(), countFunc);
+      count += std::min(minThen, minElse);
+    } else if (auto forOp = llvm::dyn_cast<scf::ForOp>(op)) {
+      count += findMinInBlock(*forOp.getBody(), countFunc);
+    } else {
+      count += countFunc(op);
+    }
+  }
+  return count;
+}
+
+// Returns the minimum found when accumulating countFunc(op) for all paths
+// between the block's start and end op
+int findMinInBlock(Block &block,
+                   const std::function<int(Operation *)> &countFunc) {
+  if (block.empty())
+    return 0;
+  return findMinBetweenOps(&block.front(), &block.back(), countFunc);
+}
+} // namespace findMin
+
+int findMinPathCountInDefChain(Value value, Operation *consumerOp,
+                               const std::function<int(Operation *)> &countFunc,
+                               int pathSum, int foundMin) {
+
+  // If the value is not defined in the same region as the consumer we need to
+  // peel the parent region of consumer until we arrive at value's region
+  while (consumerOp->getParentRegion() != value.getParentRegion()) {
+    assert(!consumerOp->getParentRegion()->empty());
+    assert(!consumerOp->getParentRegion()->front().empty());
+    pathSum += findMin::findMinBetweenOps(
```

Comment:
What if the region contains multi blocks? It would break the assumption of `findMinBetweenOps` iiuc?

Reply:
As you correctly mentioned in the other comment we want to use simply `getBlock()` here. So this is not an issue as we do not care about the region but only to paths from `consumerOp`'s block to `defValue`'s block